### ISSUE TITLE
Allow renderer to access to both actual and expected parameters

### ISSUE THREAD

**Issue by ghostbuster91** [zio/zio#2003]
Hi, I would like to enrich failure messages with more detailed information based on assertion result. This might be especially useful when dealing with large lists or just big data structures. It was previously mentioned in #1542, but probably that issue was about something else.

WDYT?

**Comment by adamgfraser** [comment]
@ghostbuster91 Sorry for the delay in getting back to you on this. Detailed failure reporting is definitely a goal we have. Can you say a little more about how you see something being reported today and how you would like to see it reported?

**Comment by ghostbuster91** [comment]
In the current implementation what is reported back to the user is the passed renderer String, but this is given upfront, before the actual evaluation of assertion, so cannot really be generated basing on the actual comparison. I would like to be able to write something like that:
```
object ExampleSpec
    extends DefaultRunnableSpec(
      suite("some suite")(
        test("failing test") {
          assert(Person(1, "asd"), MyAssertions.equalTo(Person(1, "zxc")))
        }
      )
    )

case class Person(age: Int, name: String)
```
Where inside of `MyAssertions.equalTo` I have access to both actual and expected objects before providing the assertion message. 
In other words instead of this (current implementation of equalTo):
```
final def equalTo[T](expected: T): Assertion[Any] = {
    Assertion.assertion("equalTo")(param(expected)) { actual =>
      (actual, expected) match {
        case (left, right) => left == right
      }
    }
  }
```
I  would like to be able to write:
```
  final def equalTo[T](expected: T): Assertion[Any] =
    Assertion.assertion()(param(expected)) { actual =>
      (actual, expected) match {
        case (left, right) => (left == right, "equalTo")
      }
    }
```

This would allow me to create an integration between [diffx](https://github.com/softwaremill/diffx) and zio-test

**Comment by adamgfraser** [comment]
Got it. Yes, I think that is definitely interesting. Do you want to work on this? Otherwise I would be happy to take it.

**Comment by jdegoes** [comment]
@ghostbuster91 I'd love the diffx integration!!! ❤️

**Comment by 987Nabil** [comment]
Diff messages got better and diffx is dead

**Comment by mbaechler** [comment]
Diffx is dead but also replaced by https://github.com/jatcwang/difflicious
By the way, this issue is about the ability to write or provide a user defined diff function, is it possible now?
Because otherwise I doubt ZIO reached a point where there's no need for third-party solution.
